rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,227 @@
+on Tendermint. This is done by introducing a division in the types of ways a
+network can be upgraded. There will still remain the need for hard upgrades (the
+starting of a new chain) but this will not be focused on in this
+RFC(discussions on improving this can be found [here](https://github.com/tendermint/tendermint/issues/5595)).
+Instead, this RFC will only explore what supporting soft upgrades would entail.
+
+## Proposal
+
+Tendermint supports a class of upgrades deemed as soft upgrades.
+
+Soft and hard upgrades would directly align with semver. For example, a release
+from 1.2 → 1.3 would be considered a soft upgrade and could be executed on a
+live chain (more on how that would be executed further down). A major release
+from 1.5 → 2.0 would be a hard upgrade and require a network to halt and migrate.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
+
+The block version consists of all block related data structures i.e. Header,
+Vote, Validator, Commit. This would be the sole focus for soft upgrades (a full
+list of the exact data structures can be found in
+[data structures](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md#data-structures).
+A block version could be incremented between minor releases. However, for
+backwards compatibility, all prior versions in a given major release would also
+need to be supported thus TM 1.3 could support block version 9 and 10 and TM 1.4
+could introduce block version 11 but would still need to support 9 and 10. In a
+major release the support of prior versions could be dropped. Continuing with
+the same example this could mean that TM 2.0 could support block version 11
+without supporting versions 9 or 10.
+
+The other two, P2P and App would be left as is. Any changes to the network
+protocol would require a hard upgrade and therefore would only be allowed in
+major releases. Any changes to the app version would be able to be conducted
+live and at the discretion of the app.
+
+Depending on the implementation it may be more beneficial to divide block
+protocol versions into more specific abstractions i.e. Block, State and
+Consensus to isolate changes from spilling over multiple modules (normally a
+change to say `commit` would cause us to replicate all the data structures).
+However, increasing the granularity of abstractions would compound the
+permutations between versions which could lead to untamable code, hence, given
+that these changes should be grouped together and occur seldomly it should
+suffice to have a single block protocol version.
+
+### Transitions
+
+Networks would be expected to announce upgrades in advance allowing a window for
+individual nodes to upgrade to a software version of Tendermint that would
+comply with the announced block protocol upgrade. Node runners would temporarily
+halt their node, change to the latest binary and start up again. Tendermint
```

Comment:
It might be nice to be explicit about version expectations for nodes in a network. Do we expect/require all nodes in a network to run the same minor version at all times? Patch version?

Reply:
All nodes in a network must be able to run the same protocol version (p2p, block and app). They can, however, be running different minor versions so long as that minor version supports the protocol versions

Reply:
So this means that if we upgrade Tendermint 1.4 -> 1.5, which bumps block version 10 -> 11, then all nodes in the network must upgrade to 1.5 at the same time?

Reply:
When tendermint releases 1.5, nodes can upgrade whenever they want but they will continue to run block version 10 until the entire network decides to move from 10 to 11 (this is usually done through governance proposals) at a specified height. From that height onwards nodes will switch to block version 11 unless they haven't upgraded in time in which they will error out

Reply:
Ah right, that's great. 👍

Reply:
Yeah I will try fit an example in as well. That might help explain things a bit more clearly.